File: Stage32/C/Problem9252.c
#include <stdio.h>
#include <string.h>

#define MAX 1001

char A[MAX], B[MAX];
int dp[MAX][MAX];

void lcs() {
    int n = strlen(A);
    int m = strlen(B);

    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= m; j++) {
            if (A[i - 1] == B[j - 1]) {
                dp[i][j] = dp[i - 1][j - 1] + 1;
            } else {
                dp[i][j] = (dp[i - 1][j] > dp[i][j - 1]) 
                            ? dp[i - 1][j] 
                            : dp[i][j - 1];
            }
        }
    }
}

void getLCS() {
    int i = strlen(A);
    int j = strlen(B);

    char result[MAX];
    int idx = 0;

    while (i > 0 && j > 0) {
        if (A[i - 1] == B[j - 1]) {
            result[idx++] = A[i - 1];
            i--;
            j--;
        } else {
            if (dp[i - 1][j] > dp[i][j - 1]) {
                i--;
            } else {
                j--;
            }
        }
    }

    for (int k = idx - 1; k >= 0; k--) {
        printf("%c", result[k]);
    }
    printf("\n");
}

int main() {
    scanf("%s", A);
    scanf("%s", B);

    lcs();

    printf("%d\n", dp[strlen(A)][strlen(B)]);
    getLCS();

    return 0;
}

File: Stage32/Java/Problem9252.java
package Stage32.Java;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Problem9252 {
    static String A, B;
    static int[][] dp;

    static void lcs() {
        int n = A.length();
        int m = B.length();

        dp = new int[n + 1][m + 1];

        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= m; j++) {
                if (A.charAt(i - 1) == B.charAt(j - 1)) {
                    dp[i][j] = dp[i - 1][j - 1] + 1;
                } else {
                    dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
                }
            }
        }
    }

    static String getLCS() {
        StringBuilder sb = new StringBuilder();

        int i = A.length();
        int j = B.length();

        while (i > 0 && j > 0) {
            if (A.charAt(i - 1) == B.charAt(j - 1)) {
                sb.append(A.charAt(i - 1));
                i--;
                j--;
            } else {
                if (dp[i - 1][j] > dp[i][j - 1]) {
                    i--;
                } else {
                    j--;
                }
            }
        }

        return sb.reverse().toString();
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        A = br.readLine();
        B = br.readLine();

        lcs();

        System.out.println(dp[A.length()][B.length()]);
        System.out.println(getLCS());
        br.close();
    }
}
